# <center> 3) Preparación de los Datos <center>

## Selección de Datos

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from IPython.display import display, HTML
# !(pip install chainladder)
import chainladder as cl

print("pandas: " + pd.__version__)
print("numpy: " + np.__version__)
print("chainladder: " + cl.__version__)

pandas: 1.4.2
numpy: 1.22.0
chainladder: 0.8.18


In [2]:
data=pd.read_csv('wkcomp_pos.csv')

In [17]:
data
# data.columns[4:11]

GRCODE               GRNAME  AccidentYear  DevelopmentYear  \
0          86  Allstate Ins Co Grp          1988             1988   
1          86  Allstate Ins Co Grp          1988             1989   
2          86  Allstate Ins Co Grp          1988             1990   
3          86  Allstate Ins Co Grp          1988             1991   
4          86  Allstate Ins Co Grp          1988             1992   
...       ...                  ...           ...              ...   
13195   44300   Tower Ins Co Of NY          1997             2002   
13196   44300   Tower Ins Co Of NY          1997             2003   
13197   44300   Tower Ins Co Of NY          1997             2004   
13198   44300   Tower Ins Co Of NY          1997             2005   
13199   44300   Tower Ins Co Of NY          1997             2006   

       DevelopmentLag  IncurLoss_D  CumPaidLoss_D  BulkLoss_D  \
0                   1       367404          70571      127737   
1                   2       362988         155905       60173   
2                   3       347288         220744       27763   
3                   4       330648         251595       15280   
4                   5       354690         274156       27689   
...               ...          ...            ...         ...   
13195               6          334            287          15   
13196               7          318            293           2   
13197               8          323            300           7   
13198               9          310            297           0   
13199              10          312            297           0   

       EarnedPremDIR_D  EarnedPremCeded_D  EarnedPremNet_D  Single  \
0               400699               5957           394742       0   
1               400699               5957           394742       0   
2               400699               5957           394742       0   
3               400699               5957           394742       0   
4               400699               5957           394742       0   
...                ...                ...              ...     ...   
13195             4387               3374             1013       1   
13196             4387               3374             1013       1   
13197             4387               3374             1013       1   
13198             4387               3374             1013       1   
13199             4387               3374             1013       1   

       PostedReserve97_D  
0                 281872  
1                 281872  
2                 281872  
3                 281872  
4                 281872  
...                  ...  
13195               1187  
13196               1187  
13197               1187  
13198               1187  
13199               1187  

[13200 rows x 13 columns]

### Triángulo de pérdidas

Creamos el triángulo de pérdidas para mirar como han evolucionado los pagos de las pólizas a partir del año en que ocurre el accidente, se toma como origen el año del accidente y como development el año en que se evalua el pago de la póliza. Se crean triángulos para cada una de las variables tanto de pérdidas (pólizas pagadas) como de ganancias (polizas ganadas en las compañías). El proceso se puede clasificar para cada compañía, por lo que se escoge como índice el nombre de las compañías GRNAME

In [41]:
triangle = cl.Triangle(
    data,
    origin="AccidentYear",
    development="DevelopmentYear",
    columns=data.columns[5:],
    index=["GRNAME"],
    cumulative=True,
)
triangle

Triangle Summary
Valuation:              2006-12
Grain:                     OYDY
Shape:         (132, 1, 19, 19)
Index:                 [GRNAME]
Columns:    [EarnedPremCeded_D]

In [20]:
triangle.index.head()

GRNAME
0                  Agway Ins Co
1             Alaska Nat Ins Co
2    Alaska Timber Ins Exchange
3           Allstate Ins Co Grp
4  American Contractors Ins Grp

In [21]:
print("Estructura del Chainladder triangle:", type(triangle.values))
print("Suma de los valores de la tabla de datos:", np.nansum(triangle.values))

Estructura del Chainladder triangle: <class 'numpy.ndarray'>
Suma de los valores de la tabla de datos: 768171998.0


In [22]:
triangle[[x for x in data.columns if("Loss" in x)]]

Triangle Summary
Valuation:                                   2006-12
Grain:                                          OYDY
Shape:                              (132, 3, 19, 19)
Index:                                      [GRNAME]
Columns:    [IncurLoss_D, CumPaidLoss_D, BulkLoss_D]

Frecuencias de la parte exploratoria para GRNAME, la cual nos mostraba los nombres de las 132 empresas en la base de datos.

In [23]:
freq=pd.DataFrame(data['GRNAME'].value_counts())
freq.index

Index(['Allstate Ins Co Grp', 'Bancinsure Inc', 'Harco Natl Ins Co',
       'Transportation Cas Ins Co', 'Beacon Mut Ins Co', 'Mapfre Reins Corp',
       'National American Ins Co', 'Midwest Family Mut Ins Co',
       'Associated Industries Ins Co', 'Lumbermens Underwriting Alliance',
       ...
       'Catholic Relief Ins Co Of Amer', 'Care West Ins Co', 'FFVA Mut Ins Co',
       'MO Employers Mut Ins Co', 'Martingale Natl Ins Co',
       'Hyundai Marine & Fire Ins Co Ltd', 'Agway Ins Co', 'Mada Ins Exchange',
       'Lumber Ins Cos', 'Tower Ins Co Of NY'],
      dtype='object', length=132)

Acontinuación se muestra el triángulo para de las pérdidas pagadas y gastos al final de cada año y de forma acumulada. El año de inicio es de 1988 y a lo máximo el año de evaluación en el que se ha pagado llega al año 1997, esto a pesar de que hay datos de hasta el 2006

In [24]:
triangle.loc[freq.index[0]].iloc[-1]['CumPaidLoss_D']

12        24        36        48        60        72        84        96        108       120  132  144  156  168  180  192  204  216  228
1988  70571.0  155905.0  220744.0  251595.0  274156.0  287676.0  298499.0  304873.0  321808.0  325322.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1989  66547.0  136447.0  179142.0  211343.0  231430.0  244750.0  254557.0  270059.0  273873.0  277574.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1990  52233.0  133370.0  178444.0  204442.0  222193.0  232940.0  253337.0  256788.0  261166.0  263000.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1991  59315.0  128051.0  169793.0  196685.0  213165.0  234676.0  239195.0  245499.0  247131.0  248319.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1992  39991.0   89873.0  114117.0  133003.0  154362.0  159496.0  164013.0  166212.0  167397.0  168844.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1993  19744.0   47229.0   61909.0   85099.0   87215.0   88602.0   89444.0   89899.0   90446.0   90686.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1994  20379.0   46773.0   88636.0   91077.0   92583.0   93346.0   93897.0   94165.0   94558.0   94730.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1995  18756.0   84712.0   87311.0   89200.0   90001.0   90247.0   90687.0   91068.0   91001.0   91161.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1996  42609.0   44916.0   46981.0   47899.0   48583.0   49109.0   49442.0   49073.0   49161.0   49255.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1997    691.0    2085.0    2795.0    2866.0    2905.0    2909.0    2908.0    2909.0    2909.0    2909.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1998      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1999      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2000      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2001      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2002      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2003      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2004      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2005      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2006      NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN

In [52]:
# for x in data.columns[5:]:
#     display(triangle.loc[freq.index[10]].iloc[-1][x])

### Creamos dos nuevas variables en el triángulo de pérdidas.

"CaseincurLoss": Esta nueva variable se crea restando el valor de "BulkLoss_D" de "IncurLoss_D". Básicamente, calcula la diferencia entre las pérdidas incurridas y las pérdidas a gglobal  para cada punto de datos correspondiente en el conjunto de datos.

"PaidToInc": Esta nueva variable se crea dividiendo el valor de "CumPaidLoss_D" por "IncurLoss_D". Básicamente, calcula la proporción de pérdidas incurridas que se han pagado hasta un punto específico en el tiempo.

In [44]:
triangle["CaseIncurLoss"] = triangle["IncurLoss_D"] - triangle["BulkLoss_D"]
triangle["PaidToInc"] = triangle["CumPaidLoss_D"] / triangle["IncurLoss_D"]
triangle[["CaseIncurLoss", "PaidToInc"]]

Triangle Summary
Valuation:                     2006-12
Grain:                            OYDY
Shape:                (132, 2, 19, 19)
Index:                        [GRNAME]
Columns:    [CaseIncurLoss, PaidToInc]

In [26]:
triangle["CumPaidLoss_D"].sum() / triangle["EarnedPremNet_D"].sum()

12        24        36        48        60        72        84        96        108       120  132  144  156  168  180  192  204  216  228
1988  0.169002  0.377577  0.511551  0.589170  0.641199  0.670077  0.691697  0.707762  0.726853  0.734252  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1989  0.171152  0.380515  0.510029  0.592723  0.641889  0.673263  0.695181  0.718333  0.727896  0.736776  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1990  0.170250  0.402847  0.540943  0.621712  0.669555  0.699547  0.727737  0.741724  0.753801  0.761300  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1991  0.168254  0.384413  0.510169  0.585788  0.633060  0.669131  0.685162  0.700841  0.709746  0.716366  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1992  0.156652  0.356895  0.463579  0.535713  0.585474  0.606911  0.626156  0.639385  0.648478  0.656737  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1993  0.137911  0.299451  0.386339  0.448436  0.478351  0.500951  0.514849  0.524425  0.532535  0.538605  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1994  0.135294  0.289192  0.382980  0.429647  0.460100  0.479147  0.491811  0.502014  0.509397  0.516786  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1995  0.131396  0.293704  0.367651  0.415741  0.444896  0.462494  0.475378  0.483372  0.490237  0.497043  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1996  0.157595  0.304066  0.387472  0.436638  0.467298  0.488593  0.503505  0.512192  0.520638  0.526420  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1997  0.154052  0.331010  0.426129  0.483107  0.517449  0.539827  0.554047  0.565653  0.571992  0.578609  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1998       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1999       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2000       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2001       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2002       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2003       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2004       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2005       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2006       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN

In [27]:
triangle.groupby("GRNAME").sum()

Triangle Summary
Valuation:                                            2006-12
Grain:                                                   OYDY
Shape:                                       (132, 8, 19, 19)
Index:                                               [GRNAME]
Columns:    [IncurLoss_D, CumPaidLoss_D, BulkLoss_D, Earne...

In [28]:
pd.DataFrame(np.unique(triangle["GRNAME"]))

0
0                           Agway Ins Co
1                      Alaska Nat Ins Co
2             Alaska Timber Ins Exchange
3                    Allstate Ins Co Grp
4           American Contractors Ins Grp
..                                   ...
127        Wisconsin American Mut Ins Co
128          Wisconsin Cnty Mut Ins Corp
129                    Workers Comp Exch
130  Yasuda Fire & Marine Ins Co Of Amer
131                Zurich Ins (Guam) Inc

[132 rows x 1 columns]

In [29]:
triangle.sum(axis="index").sum(axis="segments")

Triangle Summary
Valuation:                                            2006-12
Grain:                                                   OYDY
Shape:                                         (1, 8, 19, 19)
Index:                                               [GRNAME]
Columns:    [IncurLoss_D, CumPaidLoss_D, BulkLoss_D, Earne...

In [30]:
triangle.sum(axis=0).sum(axis=1)

12            24            36            48            60            72            84            96            108           120  132  144  156  168  180  192  204  216  228
1988  6.505202e+06  6.997869e+06  7.251030e+06  7.372099e+06  7.515941e+06  7.561299e+06  7.568639e+06  7.594218e+06  7.626114e+06  7.627609e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1989  6.993457e+06  7.484945e+06  7.702620e+06  7.928530e+06  8.019304e+06  8.050753e+06  8.084617e+06  8.121028e+06  8.141987e+06  8.134594e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1990  7.497081e+06  8.222617e+06  8.450322e+06  8.596685e+06  8.650997e+06  8.678112e+06  8.728115e+06  8.764842e+06  8.775955e+06  8.762355e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1991  8.307641e+06  8.883702e+06  9.128666e+06  9.234402e+06  9.288538e+06  9.360472e+06  9.398552e+06  9.424778e+06  9.414444e+06  9.406372e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1992  8.511840e+06  9.181715e+06  9.294013e+06  9.316201e+06  9.383361e+06  9.437915e+06  9.472147e+06  9.475233e+06  9.493746e+06  9.488202e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1993  9.280862e+06  9.671522e+06  9.745785e+06  9.724976e+06  9.784144e+06  9.817475e+06  9.809138e+06  9.781034e+06  9.780237e+06  9.770262e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1994  9.631596e+06  9.962505e+06  9.981473e+06  9.963320e+06  9.983721e+06  9.994179e+06  9.982154e+06  9.978186e+06  9.984483e+06  9.999210e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1995  9.607223e+06  9.916159e+06  9.952892e+06  9.865499e+06  9.889690e+06  9.871081e+06  9.880913e+06  9.884362e+06  9.913839e+06  9.943801e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1996  8.998547e+06  9.221501e+06  9.366476e+06  9.353886e+06  9.346671e+06  9.365625e+06  9.382031e+06  9.403202e+06  9.438648e+06  9.465932e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1997  8.271104e+06  8.668302e+06  8.855529e+06  8.865094e+06  8.899171e+06  8.955651e+06  8.983482e+06  9.024955e+06  9.065311e+06  9.082461e+06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1998           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1999           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2000           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2001           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2002           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2003           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2004           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2005           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2006           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN